x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java

```diff
@@ -0,0 +1,503 @@
+            final IndexRequest indexRequest = new IndexRequest(SEARCH_APPLICATION_ALIAS_NAME).opType(DocWriteRequest.OpType.INDEX)
+                .id(app.name())
+                .opType(opType)
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+                .source(buffer.bytes(), XContentType.JSON);
+            clientWithOrigin.index(indexRequest, listener);
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private void deleteSearchApplication(String resourceName, ActionListener<DeleteResponse> listener) {
+
+        try {
+            final DeleteRequest deleteRequest = new DeleteRequest(SEARCH_APPLICATION_ALIAS_NAME).id(resourceName)
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+            clientWithOrigin.delete(
+                deleteRequest,
+                new DelegatingIndexNotFoundActionListener<>(resourceName, listener, (l, deleteResponse) -> {
+                    if (deleteResponse.getResult() == DocWriteResponse.Result.NOT_FOUND) {
+                        l.onFailure(new ResourceNotFoundException(resourceName));
+                        return;
+                    }
+                    l.onResponse(deleteResponse);
+                })
+            );
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    GetAliasesResponse getAlias(String searchAliasName) {
+        return client.admin().indices().getAliases(new GetAliasesRequest(searchAliasName)).actionGet();
+    }
+
+    private void removeAlias(String searchAliasName, ActionListener<AcknowledgedResponse> listener) {
+        IndicesAliasesRequest aliasesRequest = new IndicesAliasesRequest().addAliasAction(
+            IndicesAliasesRequest.AliasActions.remove().aliases(searchAliasName).indices("*")
+        );
+        client.admin()
+            .indices()
+            .aliases(
+                aliasesRequest,
+                new DelegatingIndexNotFoundActionListener<>(
+                    searchAliasName,
+                    listener,
+                    (l, acknowledgedResponse) -> l.onResponse(AcknowledgedResponse.TRUE)
+                )
+            );
+    }
+
+    /**
+     * Deletes both the provided {@param resourceName} in the underlying index as well as the associated alias,
+     * or delegate a failure to the provided listener if the resource does not exist or failed to delete.
+     *
+     * @param resourceName The name of the {@link SearchApplication} to delete.
+     * @param listener The action listener to invoke on response/failure.
+     *
+     */
+    public void deleteSearchApplicationAndAlias(String resourceName, ActionListener<DeleteResponse> listener) {
+        removeAlias(resourceName, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                deleteSearchApplication(resourceName, listener);
+            }
```

Comment:
What happens if this is called, gets passed the `removeAlias` part, but before `deleteSearchApplication` is called, somebody calls `putSearchApplication`?

From what I can tell, it is possible to have the aliases stay around with the search application deleted.

The higher level transport calls are simply `HandledTransportActions` and even though you are using `clusterService` you could be seeing an "old" version of cluster state that is not fully committed.

I 100% expect this race condition to be hit. If we have a way out to clean it up, some warnings, etc. or if we are OK about it, please comment.

Reply:
> In any case calling deleteSearchApplication and putSearchApplication externally on the same resource without synchronisation is already a race condition?

Not that we have detected for single user usage. The Kibana UI is not performing those operations on quick succession. 

Of course multiple users could potentially run into this issue as Ben described.

Reply:
> Of course multiple users could potentially run into this issue as Ben described.

Right but then it's the same issue for index creation/deletion, mapping updates, ...

Reply:
ML has ran into the same issue multiple times. The key things are:

 - Where it is REALLY important for synchronization, use a lock (rare)
 - Where things can get out of whack, make sure its recoverable via the user (deletion, recreation, warnings, etc.)
 - Having a single thing be the "keystone" item so that failures, creations, deletions, etc. are built off its creation/deletion. In this case, it could be the SearchApplication. So, creating the alias comes AFTER creating the search application. Deleting the alias comes AFTER the search application is removed. If creating the alias fails, try not to be in a bad state, so delete the search application and indicate that creating the alias failed.

I think things that are CRUD for things that cannot have tons of race conditions should be master node operations. 

My main concern is that we think about the failure paths and bad states we can easily get into once we are making two disjoint things rely on each other.

Reply:
> Having a single thing be the "keystone" item so that failures, creations, deletions, etc. are built off its creation/deletion. In this case, it could be the SearchApplication. So, creating the alias comes AFTER creating the search application. Deleting the alias comes AFTER the search application is removed.

I like this approach. I think that's what we had first but we reordered to create/delete the alias first so that the action checks the permissions on the indices.
We could reorder and ensure that every failing paths during creation should ensure that the SearchApplication document is deleted in the system index? @carlosdelest wdyt?

Reply:
It's good to settle on always doing the SearchApplication related action first. We can add to the code paths checks to ensure Search Application doc is deleted on failure as well.

In the delete specific case, I guess we'll need to check using [has_privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-has-privileges.html) for being able to remove the alias before actually deleting the Search Application, right?

Reply:
> In the delete specific case, I guess we'll need to check using [has_privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-has-privileges.html) for being able to remove the alias before actually deleting the Search Application, right?

We could do that but I wonder if this is really needed. Could we fail the request and send back the error that only the search application was deleted?

Reply:
Sounds good, will start working on it and create a PR with the changes 👍

Reply:
@benwtrent would that alleviate your concerns here?

Reply:
@jimczi I suppose. As long as the situation is recoverable and things are clear if it causes things to break in the future (like, the user creating a search application with the same name later). 

ML actually has a nightly cleanup script to clear out orphan objects. If this gets more common, something like that here might be required.

Reply:
I've created [this draft](https://github.com/elastic/elasticsearch/pull/94924) with the changes. Looks mostly ok, but for the update case.

For updating, if we have already updated the Search App but the alias fails, then we should restore the original Search App. Which means we need to get it first. Which can go wrong.

The code is tricky already with this change (maybe that's my fault), but I think these error conditions are adding complexity to the code and there'll be multiple test cases to be done to ensure everything is working perfectly. And even so, we won't get rid of race conditions when users are interacting simultaneously.

Would another option be preferable to this? What are your thoughts?

Reply:
After a discussion with @jimczi and @ioanatia, for this PR we'd like to:

- Keep the alias creation as is and document the race condition as a known issue for a beta release
- Create a followup issue to resolve this issue, required before the feature is GA. 

Please let us know if there is any other followup work required to address this at this time. Thank you!